### PR TITLE
Scaler and descaler transformers

### PR DESCRIPTION
**Pull request by ericwayman**
**Related Issues**
Often in regression use cases the response variable is not normally distributed.  As a data scientist training regression models I'd like generic transformers to easily scale the response variable during training time, but to also descale the predictions (at scoring time) so predictions are returned in the original scale.

**Proposed Solution**
A general framework for scaling and descaling the response feature that allows developers to easily add support for new feature scaling functions with minimal boiler plate code and without having to write new code to handle reading and writing metadata.

Each family of scaling functions is represented by a case class which extends the `Scaler` case class.  
All supported families of scaling functions are stored in the `ScalingType` enum.

The `ScalerTransformer` constructs  a transformer that scales features from  `scalingType` and `scalingArgs`.  The `scalingType` and `scalingArgs` are stored in the metadata of this feature.

The `DescalerTransformer` takes in a feature for descaling and a scaled feature containing the metadata for constructing the scaling inverse.  This allows for predictions (in the scaled domain) to be descaled at scoring time.

### CODE REVIEW

File: OpPipelineStageReadWriteShared.scala [hunk 1]
```diff
@@ -0,0 +1,91 @@
+/*
```
Comment (tovbinm) [inline OpPipelineStageReadWriteShared.scala]: I thinks this file here is by mistake? you meant to update this one - https://github.com/salesforce/TransmogrifAI/blob/master/features/src/main/scala/com/salesforce/op/stages/OpPipelineStageReadWriteShared.scala ?

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/DescalerTransformerTest.scala [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import com.salesforce.op.stages.base.unary.UnaryLambdaTransformer
+import com.salesforce.op.utils.json.JsonUtils
+import com.salesforce.op.test.{TestFeatureBuilder, TestSparkContext}
+import org.joda.time.DateTime
+import org.junit.runner.RunWith
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.FlatSpec
+
+@RunWith(classOf[JUnitRunner])
+class  DescalerTransformerTest extends FlatSpec with TestSparkContext {
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/DescalerTransformerTest.scala]: please extend `OpTransformerSpec`

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/ScalerTransformerTest.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import com.salesforce.op.features.types.Real
+import com.salesforce.op.utils.json.JsonUtils
+import com.salesforce.op.test.{TestFeatureBuilder, TestSparkContext}
+import org.joda.time.DateTime
+import org.junit.runner.RunWith
+import org.scalatest.junit.JUnitRunner
+import org.scalatest.FlatSpec
+
+@RunWith(classOf[JUnitRunner])
+class  ScalerTransformerTest extends FlatSpec with TestSparkContext {
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/ScalerTransformerTest.scala]: same here - please extend `OpTransformerSpec`

File: core/src/test/scala/com/salesforce/op/stages/impl/feature/ScalerTransformerTest.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    args.slope shouldEqual 2.0
+  }
+
+  it should "Properly log scale numeric fields and serialize the transformer" in {
+    val predScaler = new ScalerTransformer[Real, Real](
+      scalingType = ScalingType.Logarithmic,
+      scalingArgs = EmptyArgs()
+    ).setInput(inA.asInstanceOf[Feature[Real]])
+    val vec: FeatureLike[Real] = predScaler.getOutput()
+    val wfModel = new OpWorkflow().setResultFeatures(vec).setInputDataset(testData).train()
```
Comment (tovbinm) [inline core/src/test/scala/com/salesforce/op/stages/impl/feature/ScalerTransformerTest.scala]: so you actually dont need a workflow to test a single transformer. see how you did it here for the date circle transformer for example ;) - https://github.com/salesforce/TransmogrifAI/blob/master/core/src/test/scala/com/salesforce/op/stages/impl/feature/DateToUnitCircleTransformerTest.scala#L84

Comment (tovbinm) [reply core/src/test/scala/com/salesforce/op/stages/impl/feature/ScalerTransformerTest.scala]: (same applies in other test cases)

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/ScalerTransformer.scala [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+/**
+ * A trait extended by a class class containing the args needed to define a family of scaling & descaling functions
+ */
+trait ScalingArgs extends JsonLike
+case class EmptyArgs() extends ScalingArgs
+case class LinearScalerArgs(slope: Double, intercept: Double) extends ScalingArgs
+
+/**
+ * Trait for defining a new family of scaling functions
+ * scalingType: a ScalingType Enum for the scaling name
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/ScalerTransformer.scala]: please move these member docs to each of the members

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DescalerTransformer.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+  def transformFn: (I1, I2) => O = (v, _) => {
+    val descaled = v.toDouble.map(scaler.descale)
+    ftFactory.newInstance(descaled)
+  }
+
+}
+
+/**
+ *  - input feature is the prediction feature to descale
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DescalerTransformer.scala]: docs

File: core/src/main/scala/com/salesforce/op/stages/impl/feature/DescalerTransformer.scala [hunk 7]
```diff
@@ -0,0 +1,78 @@
+package com.salesforce.op.stages.impl.feature
```
Comment (tovbinm) [inline core/src/main/scala/com/salesforce/op/stages/impl/feature/DescalerTransformer.scala]: license is missing

**Comment by ericwayman** [comment]
@tovbinm More tests to come

**Pull request closed without merging**